File: namastack-outbox-examples/namastack-outbox-example-java/src/main/java/io/namastack/demo/ExternalBroker.java
package io.namastack.demo;

import java.util.concurrent.ThreadLocalRandom;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ExternalBroker {

  private static final Logger logger = LoggerFactory.getLogger(ExternalBroker.class);

  private ExternalBroker() {
  }

  public static void publish(Object event, String key) {
    try {
      // Sleep between 50 and 150 milliseconds
      long sleepTime = ThreadLocalRandom.current().nextLong(50, 151);
      Thread.sleep(sleepTime);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new RuntimeException("Thread was interrupted", e);
    }

    // 40% chance of failure
    if (Math.random() < 0.4) {
      logger.warn("[External] Broker publish failed for key {}", key);
      throw new RuntimeException("Simulated failure in ExternalBroker");
    }

    logger.info("[External] Event published with key {}", key);
  }
}

File: namastack-outbox-examples/namastack-outbox-example-mariadb/src/main/kotlin/io/namastack/demo/ExternalBroker.kt
package io.namastack.demo

import org.slf4j.LoggerFactory

object ExternalBroker {
    private val logger = LoggerFactory.getLogger(ExternalBroker::class.java)

    fun publish(
        event: Any,
        key: String,
    ) {
        Thread.sleep((50..150).random().toLong())

        if (Math.random() < 0.4) { // 40% failure rate
            logger.warn("[External] Broker publish failed for key {}", key)
            throw RuntimeException("Simulated failure in ExternalBroker")
        }

        logger.info("[External] Event published with key {}", key)
    }
}
